src/libraries/System.Net.WebSockets.Client/tests/WebSocketHelper.cs

```diff
@@ -125,6 +144,98 @@ public static async Task<T> Retry<T>(ITestOutputHelper output, Func<Task<T>> fun
             }
         }
 
+        public static Task GetEchoHttp2LoopbackServer(Func<Uri, Task> clientFunc)
+        {
+            return GetEchoHttp2LoopbackServer(clientFunc, new Http2Options());
+        }
+
+        public static Task GetEchoHttp2LoopbackServer(Func<Uri, Task> clientFunc, Http2Options options)
+        {
+            Http2LoopbackServer server = Http2LoopbackServer.CreateServer(options);
+
+            Task serverTask = Task.Run(async () =>
+            {
+                var buffer = new byte[128 * 1024];
+
+                Http2LoopbackConnection connection = await server.EstablishConnectionAsync(new SettingsEntry { SettingId = SettingId.EnableConnect, Value = 1 });
+                (int streamId, HttpRequestData requestData) = await connection.ReadAndParseRequestHeaderAsync(readBody: false);
+                // send status 200 OK to establish websocket
+                await connection.SendResponseHeadersAsync(streamId, endStream: false).ConfigureAwait(false);
+
+                var webSocketStream = new Http2Stream(connection, streamId);
+                using WebSocket websocket = WebSocket.CreateFromStream(webSocketStream, true, null, TimeSpan.FromSeconds(30));
+
+                while (websocket.State == WebSocketState.Open || websocket.State == WebSocketState.CloseSent)
+                {
+                    var result = await websocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
+                    if (result.MessageType == WebSocketMessageType.Close)
+                    {
+                        if (result.CloseStatus == WebSocketCloseStatus.Empty)
+                        {
+                            await websocket.CloseAsync(WebSocketCloseStatus.Empty, null, CancellationToken.None);
+                        }
+                        else
+                        {
+                            WebSocketCloseStatus closeStatus = result.CloseStatus.GetValueOrDefault();
+                            await websocket.CloseAsync(
+                                closeStatus,
+                                result.CloseStatusDescription,
+                                CancellationToken.None);
+                        }
+
+                        continue;
+                    }
+
+                    int offset = result.Count;
+                    await websocket.SendAsync(new ArraySegment<byte>(buffer, 0, offset), result.MessageType, true, CancellationToken.None);
```

Comment:
The receive above may not get the complete message in a single chunk.
This should likely be passing `result.EndOfMessage` instead.